### PR TITLE
Add configuration for refresh token before expiry

### PR DESCRIPTION
**Pull request by jianghaolu**
Add tests later but just opening a PR for design discussion.

Fixes https://github.com/Azure/azure-sdk-for-java/issues/6182

Sample usage:
```java
ManagedIdentityCredential credential = new ManagedIdentityCredentialBuilder()
    .clientId("user managed identity client id")
    .tokenRefreshOffset(Duration.ofMinutes(10))
    .build();
```

### CODE REVIEW

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/CredentialBuilderBase.java [hunk 1]
```diff
@@ -68,4 +68,17 @@ public T httpPipeline(HttpPipeline httpPipeline) {
...
+
+    /**
+     * Sets the duration before the actual expiry of a token to refresh it.
+     * This is useful when network is congested and a request containing the
+     * token takes longer than normal to get to the server.
+     *
+     * @param refreshBeforeExpiry the duration before the actual expiry of a token to refresh it
+     */
+    @SuppressWarnings("unchecked")
+    public T setRefreshBeforeExpiry(Duration refreshBeforeExpiry) {
```
Comment (jianghaolu) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/CredentialBuilderBase.java]: A better name suggestion is welcome

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java [hunk 2]
```diff
@@ -21,6 +21,7 @@
     private Function<Duration, Duration> retryTimeout;
     private ProxyOptions proxyOptions;
     private HttpPipeline httpPipeline;
+    private Duration refreshBeforeExpiry = Duration.ofMinutes(2);
```
Comment (jianghaolu) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java]: Default matches the 2 minutes in AccessToken: https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/AccessToken.cs

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/MsalToken.java [hunk 3]
```diff
@@ -22,9 +22,10 @@
      *
      * @param msalResult the raw authentication result returned by MSAL
      */
-    public MsalToken(IAuthenticationResult msalResult) {
-        super(msalResult.accessToken(), OffsetDateTime.ofInstant(msalResult.expiresOnDate().toInstant(),
-            ZoneOffset.UTC));
+    public MsalToken(IAuthenticationResult msalResult, IdentityClientOptions options) {
+        super(msalResult.accessToken(), OffsetDateTime.ofInstant(
+                msalResult.expiresOnDate().toInstant().minus(options.getRefreshBeforeExpiry()), ZoneOffset.UTC)
+            .plusMinutes(2));
```
Comment (jianghaolu) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/MsalToken.java]: Reverts the minus 2 minutes in AccessToken: https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/core/azure-core/src/main/java/com/azure/core/credential/AccessToken.java#L22

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/CredentialBuilderBase.java (hunk 1 above)
Comment (JonathanGiles) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/CredentialBuilderBase.java]: How about `tokenRefreshOffset' or similar?

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/CredentialBuilderBase.java [hunk 4]
```diff
@@ -68,4 +68,17 @@ public T httpPipeline(HttpPipeline httpPipeline) {
         this.identityClientOptions.setHttpPipeline(httpPipeline);
         return (T) this;
     }
+
+    /**
+     * Sets the duration before the actual expiry of a token to refresh it.
+     * This is useful when network is congested and a request containing the
+     * token takes longer than normal to get to the server.
```
Comment (JonathanGiles) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/CredentialBuilderBase.java]: You should document what the default value is if it is not set.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java [hunk 5]
```diff
@@ -403,7 +401,9 @@
                             .useDelimiter("\\A");
                     String result = s.hasNext() ? s.next() : "";
 
-                    return SERIALIZER_ADAPTER.<MSIToken>deserialize(result, MSIToken.class, SerializerEncoding.JSON);
+                    MSIToken msiToken = SERIALIZER_ADAPTER.<MSIToken>deserialize(result, MSIToken.class, SerializerEncoding.JSON);
+                    return new AccessToken(msiToken.getToken(),
+                            msiToken.getExpiresAt().plusMinutes(2).minus(options.getRefreshBeforeExpiry()));
```
Comment (JonathanGiles) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java]: There is duplicate code here and above for creating the `AccessToken`, which is a possible source of bugs in the future if they are not kept in sync. Consider creating a method to centralise it.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/MsalToken.java (hunk 3 above)
Comment (JonathanGiles) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/MsalToken.java]: I feel like we are spreading the two minute concept out throughout the code base. Can you have it as a private static final value that is referred to everywhere?

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java [hunk 6]
```diff
@@ -350,7 +346,9 @@
             Scanner s = new Scanner(connection.getInputStream(), StandardCharsets.UTF_8.name()).useDelimiter("\\A");
             String result = s.hasNext() ? s.next() : "";
 
-            return Mono.just(SERIALIZER_ADAPTER.deserialize(result, MSIToken.class, SerializerEncoding.JSON));
+            MSIToken msiToken = SERIALIZER_ADAPTER.deserialize(result, MSIToken.class, SerializerEncoding.JSON);
+            return Mono.just(new AccessToken(msiToken.getToken(),
+                    msiToken.getExpiresAt().plusMinutes(2).minus(options.getRefreshBeforeExpiry())));
```
Comment (anuchandy) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java]: This adjustment of expires time looks good to me, but I have a question on when should we acquire the token from MSI endpoint. This method seems to retrieve the token as soon as it is called. Let's say it is called at 1 pm so token acquired at 1.00 pm, now let's caller subscribed to the returned Mono at 1.10 pm, now caller is served with 10 min old token. Should we delay the token acquisition until the caller subscribes to the returned mono? (I mean putting the entire acquisition logic in Mono.defer).

Comment (jianghaolu) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java]: yes I can move it inside a Mono.defer.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/CredentialBuilderBase.java (hunk 1 above)
Comment (jianghaolu) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/CredentialBuilderBase.java]: Using tokenRefreshOffset

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java (hunk 5 above)
Comment (jianghaolu) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java]: Moved to a common base class IdentityToken

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/CredentialBuilderBase.java (hunk 4 above)
Comment (jianghaolu) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/CredentialBuilderBase.java]: Documented.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java [hunk 7]
```diff
@@ -125,6 +126,29 @@ public IdentityClientOptions setHttpPipeline(HttpPipeline httpPipeline) {
...
+     * token will be considered expired at and after the time of (actual
+     * expiry - token refresh offset). The default offset is 2 minutes.
+     *
+     * This is useful when network is congested and a request containing the
+     * token takes longer than normal to get to the server.
+     *
+     * @param tokenRefreshOffset the duration before the actual expiry of a token to refresh it
+     */
+    public IdentityClientOptions setTokenRefreshOffset(Duration tokenRefreshOffset) {
+        this.tokenRefreshOffset = tokenRefreshOffset;
```
Comment (anuchandy) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java]: nit; `null` check.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java [hunk 8]
```diff
@@ -125,6 +126,29 @@ public IdentityClientOptions setHttpPipeline(HttpPipeline httpPipeline) {
         return this;
     }
 
+    /**
+<<<<<<< HEAD
```
Comment (anuchandy) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java]: nit; `<<<<<<< HEAD` from conflict resoln.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityToken.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+public class IdentityToken extends AccessToken {
+    /**
+     * Creates an identity token instance.
+     *
+     * @param token the token string.
+     * @param expiresAt the expiration time.
+     * @param options the identity client options.
+     */
+    public IdentityToken(String token, OffsetDateTime expiresAt, IdentityClientOptions options) {
+        super(token, expiresAt.plusMinutes(2).minus(options.getTokenRefreshOffset()));
```
Comment (anuchandy) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityToken.java]: nit; `Object.requireNonNull` checks.

Comment (anuchandy) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityToken.java]: discussed with Jinaghao offline, these are impl classes and these values are derived from service returned response. Hence we don't have to check here given we shouldn't test service behavior.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java (hunk 8 above)
Comment (jianghaolu) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java]: Removed.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java (hunk 7 above)
Comment (jianghaolu) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClientOptions.java]: Added a nullcheck

**Pull request merged**